crates/bevy_tasks/src/iter/mod.rs

```diff
@@ -0,0 +1,444 @@
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        let mut last_item = None;
+        while let Some(batch) = self.next_batch() {
+            last_item = batch.last();
+        }
+        last_item
+    }
+
+    // TODO: Optimize with size_hint on each batch
+    fn nth(mut self, n: usize) -> Option<Self::Item> {
+        let mut i = 0;
+        while let Some(batch) = self.next_batch() {
+            for item in batch {
+                if i == n {
+                    return Some(item);
+                }
+                i += 1;
+            }
+        }
+        None
+    }
+
+    // TODO: Use IntoParallelIterator for U
+    fn chain<U>(self, other: U) -> Chain<Self, U>
+    where
+        U: ParallelIterator<B, Item = Self::Item>,
+    {
+        Chain {
+            left: self,
+            right: other,
+            left_in_progress: true,
+        }
+    }
+
+    fn map<T, F>(self, f: F) -> Map<Self, F>
+    where
+        F: FnMut(Self::Item) -> T + Send + Clone,
+    {
+        Map { iter: self, f }
+    }
+
+    fn for_each<F>(mut self, pool: &TaskPool, f: F)
+    where
+        F: FnMut(Self::Item) + Send + Clone + Sync,
+    {
+        pool.scope(|s| {
+            while let Some(batch) = self.next_batch() {
+                let newf = f.clone();
+                s.spawn(async move {
+                    batch.for_each(newf);
+                });
+            }
+        });
+    }
+
+    fn filter<F>(self, predicate: F) -> Filter<Self, F>
+    where
+        F: FnMut(&Self::Item) -> bool,
+    {
+        Filter {
+            iter: self,
+            predicate,
+        }
+    }
+
+    fn filter_map<R, F>(self, f: F) -> FilterMap<Self, F>
+    where
+        F: FnMut(Self::Item) -> Option<R>,
+    {
+        FilterMap { iter: self, f }
+    }
+
+    fn flat_map<U, F>(self, f: F) -> FlatMap<Self, F>
+    where
+        F: FnMut(Self::Item) -> U,
+        U: IntoIterator,
+    {
+        FlatMap { iter: self, f }
+    }
+
+    fn flatten(self) -> Flatten<Self>
+    where
+        Self::Item: IntoIterator,
+    {
+        Flatten { iter: self }
+    }
+
+    fn fuse(self) -> Fuse<Self> {
+        Fuse { iter: Some(self) }
+    }
+
+    fn inspect<F>(self, f: F) -> Inspect<Self, F>
+    where
+        F: FnMut(&Self::Item),
+    {
+        Inspect { iter: self, f }
+    }
+
+    fn by_ref(&mut self) -> &mut Self {
+        self
+    }
+
+    // TODO: Investigate optimizations for less copying
+    fn collect<C>(mut self, pool: &TaskPool) -> C
+    where
+        C: std::iter::FromIterator<Self::Item>,
+        Self::Item: Send + 'static,
+    {
+        pool.scope(|s| {
+            while let Some(batch) = self.next_batch() {
+                s.spawn(async move { batch.collect::<Vec<_>>() });
+            }
+        })
+        .into_iter()
+        .flatten()
+        .collect()
+    }
+
+    // TODO: Investigate optimizations for less copying
+    fn partition<C, F>(mut self, pool: &TaskPool, f: F) -> (C, C)
```

Comment:
I definitely understand not wanting to duplicate the docs for the standard library, but a very short description for what it does and a link to the standard docs would be helpful. In fact, I think it's good to say a bit less so that the implementation-specific notes (for example all() not short-circuiting) are more likely to be seen

Reply:
One potential improvement would be to partition by batch instead of item and return an iterator for each set of batches. (I feel like it's something that could be done in a follow-up change although it does change behavior in a non-trivial way.)

Reply:
I've added short decriptions of each ParallelIterator method and linked each to the relevant Iterator docs.

For partition, I'd rather leave performance improvements for the future. I think small breaking changes are still acceptable.